### PR TITLE
#3003 PermissionReducer and permission changes

### LINKED ISSUE #3003: Check for location permission before temperature syncing
## Is your feature request related to a problem? Please describe.

In order to use bluetooth functionalioty, the app must have the access to the devices location.

## Describe the solution you'd like

- Need to check for devices location permission. Currently we check for bluetooth permission in the native module. It's better for this functionality to not be in the bluetooth module. So, shift bluetooth/location permission requests to the action which starts a sync.
- would be good to make some UI decisions - i.e. sync disabled - so i think: `PermissionReducer`

## Implementation

- Create a `PermissionReducer`
- Create actions `PermissionActions` to try get permission for bluetooth and location
- Call permission actions from temperature syncing actions

## Describe alternatives you've considered

N/A

## Additional context

N/A

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #3003 

## Change summary

**Permission stuff is in the docs here: https://docs.google.com/document/d/1UPa2HNmUhithGoK0anmrQAKxkxcHpnJqG6W2H6sQZD4/edit#

- Added a `PermissionReducer`, `PermissionActions`, `PermissionSelectors` to manage permission state within the app so that rendering can be dependent on stuff - using the app state coming into the foreground to recheck permissions (for when they're enabled outside of the app).
- Simplified the exporting permission request.
- Added functionality for temperature syncing:
1. Bluetooth will just be enabled, if possible. Otherwise, if not (i.e. the tablet has forced it off somehow or doesn't have it)
Screen Shot 2020-07-03 at 11 52 57 AM
2. When a sync occurs passively/manually, and bluetooth can't be enabled, a toast will pop up:
image
3. The location permission needs to be got to enabled bluetooth usage (why? I dunno, ask google!). When a sync occurs through a MANUAL sync this comes up:
image
4. If they decline they'll get a toast and the sync will be disabled
image
5. The state will show disabled here:
Screen Shot 2020-07-03 at 11 53 28 AM

## Testing

N/A

### Related areas to think about

I gots a problem here (overlapping text/states/msupply man): 
Screen Shot 2020-07-03 at 12 12 39 PM

Any ideas? :(

### CODE REVIEW

File: src/mSupplyMobileApp.js [hunk 1]
```diff
@@ -84,6 +85,10 @@ class MSupplyMobileAppContainer extends React.Component {
   }
 
   componentDidMount = () => {
+    const { dispatch } = this.props;
+
+    dispatch(PermissionActions.checkPermissions());
```
Comment (joshxg) [inline src/mSupplyMobileApp.js:94]: when we open the app, check our permission situation

File: src/mSupplyMobileApp.js [hunk 2]
```diff
@@ -102,6 +107,7 @@ class MSupplyMobileAppContainer extends React.Component {
     const { dispatch } = this.props;
     if (nextAppState?.match(/inactive|background/)) dispatch(UserActions.setTime());
     if (appState?.match(/inactive|background/) && nextAppState === 'active') {
+      dispatch(PermissionActions.checkPermissions());
```
Comment (joshxg) [inline src/mSupplyMobileApp.js:115]: when coming into the foreground, check permissions again!

File: src/mSupplyMobileApp.js [hunk 3]
```diff
@@ -84,12 +86,20 @@ class MSupplyMobileAppContainer extends React.Component {
   }
 
   componentDidMount = () => {
+    const { dispatch } = this.props;
+
+    BluetoothStatus.addListener(newStatus =>
+      dispatch(PermissionActions.requestBluetooth(newStatus))
+    );
```
Comment (joshxg) [inline src/mSupplyMobileApp.js:93]: dispatch the state change action when bluetooth changes

File: src/mSupplyMobileApp.js [hunk 4]
```diff
@@ -84,12 +86,20 @@ class MSupplyMobileAppContainer extends React.Component {
...
+    dispatch(PermissionActions.checkPermissions());
+
     if (!__DEV__) {
       AppState.addEventListener('change', this.onAppStateChange);
     }
   };
 
   componentWillUnmount = () => {
+    BluetoothStatus.removeListener();
```
Comment (joshxg) [inline src/mSupplyMobileApp.js:102]: only one listener for this one

**Comment by mark-prins** [comment]
> Any ideas? :(

- move the sensor sync to the left hand side of the screen
- change the message "sync error" and "sync disabled" and "sync ok" to icons ( such as the material UI version below )
Screen Shot 2020-07-03 at 2 25 47 PM
- a *really* small font?

**Comment by joshxg** [comment]
@mark-prins 

- Nice. On the left - usually theres a page title.
- I like changing the icon - do you mean change the current icon to this smaller one, or change the font to the icon?

File: src/selectors/temperatureSync.js [hunk 5]
```diff
@@ -31,8 +33,9 @@ export const selectTemperatureSyncMessage = ({ temperatureSync }) => {
 
 export const selectTemperatureSyncLastSyncString = ({ temperatureSync }) => {
   const { lastTemperatureSync } = temperatureSync;
+  const asMoment = moment(lastTemperatureSync);
 
-  return `${moment(lastTemperatureSync).format('H:mm MMMM D, YYYY ')}`;
+  return `${asMoment.isValid() ? asMoment.format('H:mm MMMM D, YYYY ') : '--'}`;
```
Comment (mark-prins) [inline src/selectors/temperatureSync.js:38]: nice!

File: src/localization/syncStrings.json [hunk 6]
```diff
@@ -22,7 +23,11 @@
     "error_resetting_sensor": "Error resetting sensor",
     "saving_temperature_logs": "Saving temperature logs",
     "no_sensors": "No sensors to scan for",
-    "syncing_temperatures": "Syncing temperatures"
+    "syncing_temperatures": "Syncing temperatures",
+    "bluetooth_disabled": "Sync error. Bluetooth disabled.",
+    "please_enable_bluetooth": "We tried syncing temperatures, but your bluetooth is diabled. Enabled it to sync temperatures!",
+    "location_permission": "Sync error. Location permission required.",
+    "grant_location_permission": "We tried syncing temperatures, but we don't have access to your location. Grant us permission so we can sync temperatures!"
```
[the remark this answers is not in the document]
Comment (mark-prins) [reply src/localization/syncStrings.json:30]: They read fine to me. Ideally translate why 'location' is needed to sync bluetooth.. but as I have no idea why, can't help with that 🤷‍♂️

**Comment by mark-prins** [comment]
> I like changing the icon - do you mean change the current icon to this smaller one, or change the font to the icon?

I was originally thinking to change the current arrows:

Screen Shot 2020-07-03 at 2 51 23 PM

and stick a diagonal line through to indicate 'disabled'. With maybe a red/green and exclamation mark indicating the 'broken' state.

I'm now thinking to replace them with the curved arrows, and use the three different versions - they're fairly clear as to the state of things?

**Comment by joshxg** [comment]
Yeah - sounds good to me! Cheers

**Comment by joshxg** [comment]
Thanks @mark-prins have added the icon changes 

Normal sync with an error to be as long as pos!

With errors:
image

When disabled: 
image

When enabled and all OK
image


Think it looks much better - although still concerned around the affordance, it does look much better!

**Comment by joshxg** [comment]
whoops, lucky I put the pics! missing a fullstop in the pics that I've put back on `sync enabled` 🤣

**Pull request merged**